components/sidebar/sidebar_category/sidebar_category_sorting_menu/sidebar_category_sorting_menu.tsx

```diff
@@ -0,0 +1,213 @@
+}
+
+export class SidebarCategorySortingMenu extends React.PureComponent<Props, State> {
+    isLeaving: boolean;
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            openUp: false,
+            width: 0,
+            selectedDmNumber: 20,
+            selectedCategorySort: CategorySorting.Alphabetical,
+        };
+
+        this.isLeaving = false;
+    }
+
+    handleSortDirectMessages = (sorting: CategorySorting) => {
+        const {category} = this.props;
+
+        this.props.actions.setCategorySorting(category.id, sorting);
+        trackEvent('ui', `ui_sidebar_sort_dm_${sorting}`);
+        this.setState({selectedCategorySort: sorting});
+    }
+
+    handlelimitVisibleDMsGMs = (number: number) => {
+        const {currentUserId} = this.props;
+        this.props.actions.savePreferences(currentUserId, [{
+            user_id: currentUserId,
+            category: Constants.Preferences.CATEGORY_SIDEBAR_SETTINGS,
+            name: Settings.limitVisibleDmsGms,
+            value: `${number}`,
+        }]);
+        this.setState({selectedDmNumber: number});
+    }
+
+    handleOpenDirectMessagesModal = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+        e.stopPropagation();
+        this.props.handleOpenMoreDirectChannelsModal(e.nativeEvent);
+        trackEvent('ui', 'ui_sidebar_create_direct_message');
+    }
+
+    renderDropdownItems = () => {
+        const {intl} = this.props;
+        const {selectedDmNumber, selectedCategorySort} = this.state;
+
+        const sortMenuItems: SubmenuItemProps[] = [{
+            id: 'sortAlphabetical',
+            direction: 'right' as any,
+            text: intl.formatMessage({id: 'user.settings.sidebar.sortAlpha', defaultMessage: 'Alphabetically'}),
+            action: () => this.handleSortDirectMessages(CategorySorting.Alphabetical),
+        },
+        {
+            id: 'sortByMostRecent',
+            direction: 'right' as any,
+            text: intl.formatMessage({id: 'sidebar.sortedByRecencyLabel', defaultMessage: 'Recent Activity'}),
+            action: () => this.handleSortDirectMessages(CategorySorting.Recency),
+        }];
+
+        const dmLimitOptions = [10, 15, 20, 40];
+
+        const selectedDmCount = dmLimitOptions.map((number) => {
+            return {
+                id: `SidebarCategorySortingMenu-dmCount-${number}`,
+                direction: 'right' as any,
+                text: number,
+                action: () => this.handlelimitVisibleDMsGMs(number),
+            } as SubmenuItemProps;
+        });
+
+        const categoryMenuItems: SubmenuItemProps[] = [];
+        categoryMenuItems.push(
+            {
+                id: 'showAllDms',
+                direction: 'right' as any,
+                text: intl.formatMessage({id: 'sidebar.allDirectMessages', defaultMessage: 'All direct messages'}),
+                action: () => this.handlelimitVisibleDMsGMs(Infinity),
```

Comment:
I think we might want to use 0 or -1 for this instead of infinity. Infinity is technically correct for this and it makes the selector slightly simpler since you can just pass it right into slice, but I could see there being difficulties since storing the string "infinity" feels weird.

Reply:
If we pass 0 or -1 or even empty string or null it will remove all items from the array and that adds an extra check for the selector if the value === 0, 3/5 this is more simplified IMO

Reply:
My concern is that Infinity isn't a valid integer, so we're somewhat mixing types with the value stored for this. I just checked and storing `Infinity.toString()` in a preference works fine, but then when we pull it back out of preferences `parseInt('Infinity')` returns NaN instead of infinty, so we might end up with problems because we're working outside of well-defined numbers.